### DISCUSSION TITLE
Invalid sourceUrl for some NuGet dependencies affecting monorepos

### DISCUSSION THREAD

**Discussion by Turnerj** [renovatebot/renovate#22887] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.133.3

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

There is an issue around `sourceUrl` handling for NuGet packages where older versions of the package don't have a repository URL defined which breaks handling of monorepos.

For example, take the package `Microsoft.EntityFrameworkCore`. If I am upgrading from 6.0.16 to 6.0.18, I'm expecting the source URL to be `https://github.com/dotnet/efcore`  but I get `http://www.asp.net/`.

This is strange because versions 6.0.16, 6.0.18 (latest v6) and 7.0.7 (the actual latest stable) all have `https://github.com/dotnet/efcore` as the repository URL which would become the `sourceUrl` for Renovate.
- https://api.nuget.org/v3-flatcontainer/microsoft.entityframeworkcore/6.0.16/microsoft.entityframeworkcore.nuspec
- https://api.nuget.org/v3-flatcontainer/microsoft.entityframeworkcore/6.0.18/microsoft.entityframeworkcore.nuspec
- https://api.nuget.org/v3-flatcontainer/microsoft.entityframeworkcore/7.0.7/microsoft.entityframeworkcore.nuspec

What I found though is that version 1.0.0 of `Microsoft.EntityFrameworkCore` doesn't have the repository URL defined but does have a project URL specifically with `http://www.asp.net/`.
- https://api.nuget.org/v3-flatcontainer/microsoft.entityframeworkcore/1.0.0/microsoft.entityframeworkcore.nuspec

What I think is the problem is this block of code:

https://github.com/renovatebot/renovate/blob/a95d8a136d2e2e4c8a964836a105229d17554ef1/lib/modules/datasource/nuget/v3.ts#L128-L143

Specifically, that the setting of `latestStable` assumes the order of versions starts from the oldest to the newest but there is no guarantee of that. I've so far been sharing links to the official NuGet API however I'm actually using an Azure DevOps artifacts feed that has NuGet as an upstream provider which actually has the order from newest to oldest (which I have no control of). When the Renovate code tries to find the `latestStable`, it ends up picking the oldest stable release causing the problem.

This doesn't affect every dependency as part of the dotnet monorepo as the oldest version of some packages actually have a repository URL defined so the dotnet monorepo still works correctly.

What I think Renovate should be doing though is not assuming the ordering (unless there is a NuGet API spec that says one way or another) and instead sort by the parsed semantic versions provided by the API instead.

### Relevant debug logs

<details><summary>Logs</summary>

```
...

               {
                 "datasource": "nuget",
                 "depType": "nuget",
                 "depName": "Microsoft.EntityFrameworkCore",
                 "currentValue": "6.0.16",
                 "registryUrls": [
                   "https://pkgs.dev.azure.com/xxxxx/_packaging/xxxxx/nuget/v3/index.json#protocolVersion=3",
                   "https://pkgs.dev.azure.com/xxxxx/_packaging/yyyyy/nuget/v3/index.json#protocolVersion=3",
                   "https://api.nuget.org/v3/index.json#protocolVersion=3"
                 ],
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "6.0.18",
                     "newValue": "6.0.18",
                     "releaseTimestamp": "2023-06-14T06:00:49.000Z",
                     "registryUrl": "https://pkgs.dev.azure.com/xxxxx/_packaging/xxxxx/nuget/v3/index.json#protocolVersion=3",
                     "newMajor": 6,
                     "newMinor": 0,
                     "updateType": "patch",
                     "branchName": "renovate/microsoft.entityframeworkcore-6.x"
                   },
                   {
                     "bucket": "major",
                     "newVersion": "7.0.7",
                     "newValue": "7.0.7",
                     "releaseTimestamp": "2023-06-13T14:05:18.000Z",
                     "registryUrl": "https://pkgs.dev.azure.com/xxxxx/_packaging/xxxxx/nuget/v3/index.json#protocolVersion=3",
                     "newMajor": 7,
                     "newMinor": 0,
                     "updateType": "major",
                     "branchName": "renovate/microsoft.entityframeworkcore-7.x"
                   }
                 ],
                 "packageName": "Microsoft.EntityFrameworkCore",
                 "warnings": [],
                 "versioning": "nuget",
                 "sourceUrl": "http://www.asp.net/",
                 "homepage": "http://www.asp.net/",
                 "currentVersion": "6.0.16",
                 "isSingleVersion": true,
                 "fixedVersion": "6.0.16"
               },

...

DEBUG: Found sourceUrl with multiple branches that should probably be combined into a group (repository=xxxxx, baseBranch=develop)
       "sourceUrl": "http://www.asp.net/",
       "newVersion": "6.0.18",
       "branches": {
         "renovate/microsoft.entityframeworkcore.design-6.x": "Microsoft.EntityFrameworkCore.Design",
         "renovate/microsoft.entityframeworkcore.sqlserver-6.x": "Microsoft.EntityFrameworkCore.SQLServer",
         "renovate/microsoft.entityframeworkcore.relational-6.x": "Microsoft.EntityFrameworkCore.Relational",
         "renovate/microsoft.entityframeworkcore.inmemory-6.x": "Microsoft.EntityFrameworkCore.InMemory",
         "renovate/microsoft.entityframeworkcore-6.x": "Microsoft.EntityFrameworkCore",
         "renovate/microsoft.aspnetcore.authentication.openidconnect-6.x": "Microsoft.AspNetCore.Authentication.OpenIdConnect",
         "renovate/microsoft.entityframeworkcore.sqlite-6.x": "Microsoft.EntityFrameworkCore.Sqlite",
         "renovate/microsoft.data.sqlite-6.x": "Microsoft.Data.Sqlite",
         "renovate/microsoft.aspnetcore.middlewareanalysis-6.x": "Microsoft.AspNetCore.MiddlewareAnalysis",
         "renovate/microsoft.aspnetcore.jsonpatch-6.x": "Microsoft.AspNetCore.JsonPatch"
       }

```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by chrisvanderpennen** [comment]
We're seeing this because we're using an Azure DevOps package feed as a pull-through cache of nuget.org.  Not sure if or when they changed the API sort order to be semver descending instead of ascending, but so far as I can tell from [the API documentation](https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource#registration-page-object) the ordering of the `.items` array is unspecified.

On the bright side, from my experimentation this looks like a very simple fix:
```typescript
if (
  semver.valid(version) &&
  !semver.prerelease(version) &&
  (latestStable === null || semver.compare(latestStable, version) < 0)
) {
  latestStable = removeBuildMeta(version);
  homepage = projectUrl ? massageUrl(projectUrl) : homepage;
}
```

I'd be happy to get a PR up including a unit test against a recorded Azure DevOps API response if there's interest.

**Reply by viceice** [reply]
please do 🙏

**Reply by chrisvanderpennen** [reply]
I found some edge cases around pre-release versions and NuGet's almost-semver xx.yy.zz.aa format.  I've tried a few different approaches to resolving this with those in mind, and the simplest solution I've arrived at is to use the [NugetVersioningApi](https://github.com/renovatebot/renovate/blob/main/lib/modules/versioning/nuget/index.ts) module to sort and process the API response.

I'm pretty confident it's safe to switch from plain semver to NugetVersioningApi because [the NuGet versioning spec](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning) is quite strict and enforced by the NuGet API.  If Renovate is configured to use a different version strategy for a particular NuGet package, the versions will be re-sorted using the configured strategy later by the main datasource module.  The NuGet datasource currently assumes that versions are both sorted by the API and valid semver anyway, so I don't think this will harm overall behaviour.

The overall change becomes:
```typescript
import { api as versioning } from '../../versioning/nuget';

...

const catalogEntries = (await p.all(catalogPagesQueue))
  .flat()
  .sort((a, b) => versioning.sortVersions(a.version, b.version));

...

if (versioning.isValid(version) && versioning.isStable(version)) {
  latestStable = removeBuildMeta(version);
  homepage = projectUrl ? massageUrl(projectUrl) : homepage;
}
```

The Maven datasource already does this, so I think it's reasonable to do here as well.  If you have any suggestions or concerns please let me know, or I can open a draft PR and we can discuss there.
https://github.com/renovatebot/renovate/blob/2d77984e6a295bbf9b1861b5532805d8ba793656/lib/modules/datasource/maven/index.ts#L9-L11
https://github.com/renovatebot/renovate/blob/2d77984e6a295bbf9b1861b5532805d8ba793656/lib/modules/datasource/maven/index.ts#L31-L44

**Reply by viceice** [reply]
please open a PR
